# Tutorial: Detection of Algorithmic and Data Biases in Integrated Single-Cell RNA-Sequencing Datasets
This package aims to detect batch effects from batch-corrected integrated scRNA-seq datasets and to automaticlly find the best integration methods using Machine Learning Auditors.

The tutorial is arranged as follows:
1. Dependency requirements
2. Data Preparation
3. Test demo

## Dependency requirements
Begin by clone the repository. and install dependencies.
```
git clone https://github.com/SheltonZhaoK/Detection-of-Algorithmic-and-Data-Biases-in-Integrated-Single-Cell-RNA-Sequencing-Datasets.git
```
The scripts are written in Python. A virtual python environment can be created with following commands:
``` 
python3 -m venv <dir>
```
The python environment could be activated by:
```
source <dir>/bin/activate
```
The python environment could be deactivated by:
```
deactivate
```
After activating the python virtual environment, all dependencies could be installed by:
```
pip3 install -r requirements.txt
```

